### ISSUE TITLE
Session not available when running with Jetty 9.4

### ISSUE THREAD

**Issue by Artur-** [Atmosphere/atmosphere#2241]
Using Jetty 9.4.0 causes session variables set in the initial HTTP request not being available in the websocket request. This shows the same symptoms as #2157 but using Jetty 9.4 and JSR356. The test case is exactly the same, available at https://github.com/Artur-/atmosphere-samples/commit/567db9c0d17a5ef6be7b5ec88216155d7521399f

Reported in https://github.com/vaadin/framework/issues/8134

**Comment by Artur-** [comment]
Is this a Jetty issue or an Atmosphere issue?

**Comment by jfarcand** [comment]
It's more because they have changed (again) their internal API.

**Comment by Artur-** [comment]
Wrong issue? This is about JSR356

**Comment by trivore-kari** [comment]
Artur, we (too) have triaged this, and Jetty really has changed their api in dotdot release and that breaks Atmosphere and then Vaadin.

**Comment by jfarcand** [comment]
Should be fixed by by #2270

**Comment by Artur-** [comment]
Still the same problem with #2270 (#2270 fixes Jetty 9.3.15 compatibility but not 9.4 it seems)

**Comment by Artur-** [comment]
The problem comes from 
```
                    .session((HttpSession) handshakeRequest.getHttpSession())
```
at https://github.com/Atmosphere/atmosphere/blob/master/modules/runtime/src/main/java/org/atmosphere/container/JSR356Endpoint.java#L219

where `handshakeRequest.getHttpSession()` returns null with Jetty 9.4.x and the HTTP session with Jetty 9.3.x and earlier

**Comment by Artur-** [comment]
Maybe @joakime could comment if this is a Jetty or Atmosphere problem

**Comment by joakime** [comment]
@trivore-kari  know that jetty versioning is `<servlet_support>.<major>.<minor>` (been this way since 1996)

Per spec, the HttpServletRequest, HttpServletResponse, HttpSession, and all things that are part of the HTTP phase of the WebSocket are recycled once the HTTP Upgrade is complete (when the HTTP Upgrade Handshake completed, you are no longer HTTP, you are whatever protocol that upgrade resulted in be it WebSocket, HTTP/2, stunnel, etc ...)

Once the HTTP Upgrade is complete, the javax.websocket.server.HandshakeRequest.getHttpSession() can return a null HttpSession (as it might have been recycled).

Also of note (this is still an open issue for JSR356, CDI, and the Servlet specs btw) HttpSession information outside of the actual Handshake is undefined, the information is not live, the information is unconnected to its storage, the object lifecycle is undefined, the lifespan of the information is undefined, the HttpSession object itself is a ghost and cannot be trusted.

It is recommended by all groups (not just Jetty) that if you have any information you need from the HttpServletRequest, HttpServletResponse, or HttpSession you copy that into your own locations at Upgrade Handshake time. (this includes using features like Request.isUserInRole(), Request.setAttribute(), etc)

An example highlighting this undetermined behavior/state ...

1. Normal HTTP GET request, HttpSession created, foo=bar added
2. WebSocket Upgrade performed
3. Another HTTP GET request that updates the value 'foo=fleb' in the HttpSession
4. WebSocket accesses the value for 'foo' in the HttpSession and gets the old value 'bar'

Another example

1. Normal HTTP GET request, HttpSession created, foo=bar added.
2. WebSocket Upgrade performed
3. Another HTTP GET request that invalidates the HttpSession (logout for example)
4. WebSocket accesses the value for 'foo' in HttpSession and gets 'bar'

Another example

1. Normal HTTP GET request, HttpSession created, 'foo=bar' added.
2. WebSocket Upgrade performed
3. WebSocket set the value for 'foo' in HttpSession to 'foo=huh'
4. Another HTTP GET request that requests the value of 'foo' sees the answer 'bar'

Another example

1. Normal HTTP GET request, HttpSession created (expires in 10 minutes), 'foo=bar' added.
2. WebSocket Upgrade performed
3. (30 minutes later) WebSocket access 'foo' in the HttpSession and gets 'bar'

Servlet 4.0 has not fixed these open issues with JSR356, or any post-Upgrade Lifecycle of objects.
The CDI project has also not address the post-Upgrade LifeCycle behavior (they are waiting on either the JSR356 update or the Servlet updates specifying this behavior).

Ok, with the specs out of the way...

* [Jetty's javax.websocket.server.HandshakeRequest.getHttpSession()](https://github.com/eclipse/jetty.project/blob/jetty-9.4.5.v20170502/jetty-websocket/javax-websocket-server-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/server/JsrHandshakeRequest.java#L47-L50) implementation
* [Jetty's ServletUpgradeRequest.getSession()](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/ServletUpgradeRequest.java#L371-L381) implementation
* [Jetty's UpgradeHttpServletRequest.getSession(boolean)](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/UpgradeHttpServletRequest.java#L262-L269) implementation

What this means is that at the time your are calling `javax.websocket.server.HandshakeRequest.getSession()` jetty asks the `HttpServletRequest.getHttpSession(false)` for it.

The `HttpServletRequest` object itself could be recycled already (if you are attempting to access it from outside of the Upgrade Handshake phase)

If the `HttpServletRequest` object is not recycled, then it could be that there is no information in the request referencing a real, created prior, `HttpSession`, so a new one will not be created (It cannot create the HttpSession, per the Servlet Spec the HTTP response no longer exists, and in order to create an HttpSession it needs to include information in the HTTP response indicating that, such as a Set-Cookie header)

You might be asking why the HttpSession object doesn't just exist in the javax.websocket.server.HandshakeRequest?  This is because the Servlet spec says the HttpSession object only exists in the HttpServletRequest object once its attempted to be accessed via the `.getHttpSession()` calls.   Depending on server configuration, creating an HttpSession can be an extremely heavyweight (either in terms of time, or resources), so its only created when accessed.

Also, creating a fresh HttpSession (for the first time) at the time of WebSocket handshake upgrade is actually not supported by many server configurations and/or user-agents, as the configured means to report that session reference back to the user-agent might not be available in that situation.

In short, if you have something in the HttpSession that you need, get it out from the HttpSession at Upgrade Handshake time, not later.

**Comment by trivore-kari** [comment]
@joakime, thank you for that info on version numbering. Very logical, actually. News to me. My bad, I did not research this. (And thanks for Jetty, our developers love it :-)

**Comment by Artur-** [comment]
Thanks for the detailed answer. Atmosphere does indeed keep a reference to the handshake request until `EndPoint.onOpen` is triggered. This seems to be avoidable by instead storing the needed parts of the request (some headers and the session reference). Not sure how "safe" this is but it seems to work also for Jetty 9.4 at least in the Vaadin case

**Comment by jfarcand** [comment]
Thanks @Artur- and big @joakime for the explanation (maybe I should follow more jsr356 spec list :)). 

Included in `2.4.12`
